app/controllers/katello/api/v2/subscriptions_controller.rb

```diff
@@ -200,6 +199,11 @@ def available_for_activation_key
       @activation_key.available_subscriptions
     end
 
+    def subscription_not_relevant?
```

Comment:
There are a *lot* of negations going on in this code and it's hurting my brain. 💀 

One suggestion is to frame this method positively ie `subscription_relevant?` and rework the logic accordingly

However, I think it could be clarified significantly doing something like this:

```

def permitted_org_ids
  [User.current.organizations&.pluck(:id), Organization.current&.id].flatten.uniq.compact
end


unless User.current.admin? || !permitted_org_ids.include?(@resource.organization_id)
  # raise error
end
```

Now I can read the logic the same way I would describe the intended behavior: unless the user is an admin or subscription's org is permitted then raise an error.

The conditional could even be written to avoid the negation: `unless User.current.admin || permitted_org_ids.exclude?(@resource.organization_id)`

Reply:
I was thinking about this some more and thought there could be a way to use our permissions system to handle this rather than doing this check in the controller. Your controller tests pass with this:

controller:

```
unless @resource.readable?
  # raise error
end
```

app/models/katello/authorization/subscription.rb
```
module ClassMethods
  def readable
    return all if User.current.admin?

    authorized(:view_subscriptions).where(organization_id: User.current.organization_ids)
  end
end
```

app/models/katello/authorization/pool.rb
```
    def readable?
      self.class.readable.where(id: self.id).any?
    end

    module ClassMethods
      def readable
        where(:subscription_id => Katello::Subscription.readable)
      end
    end
```

We end up with more reusable code, and all the checks are done at the db via 3 queries:

```
Katello::Pool Load (0.3ms)  SELECT  "katello_pools".* FROM "katello_pools" WHERE (katello_pools.id = (128914721) or katello_pools.cp_id = ('128914721')) ORDER BY "katello_pools"."id" ASC LIMIT $1  [["LIMIT", 1]]

Filter Load (2.1ms)  SELECT DISTINCT "filters".* FROM "filters" INNER JOIN "filterings" ON "filterings"."filter_id" = "filters"."id" INNER JOIN "permissions" ON "permissions"."id" = "filterings"."permission_id" INNER JOIN "roles" ON "filters"."role_id" = "roles"."id" INNER JOIN "cached_user_roles" ON "roles"."id" = "cached_user_roles"."role_id" LEFT JOIN taxable_taxonomies ON (filters.id = taxable_taxonomies.taxable_id AND taxable_type = 'Filter') LEFT JOIN taxonomies ON (taxonomies.id = taxable_taxonomies.taxonomy_id) WHERE "cached_user_roles"."user_id" = $1 AND (permissions.resource_type = 'Katello::Subscription') AND (permissions.name = 'view_subscriptions') AND (taxable_taxonomies.id IS NULL OR (taxonomies.type = 'Organization') OR (taxonomies.type = 'Location'))  [["user_id", 980190962]]

Katello::Subscription Exists (0.3ms)  SELECT  1 AS one FROM "katello_subscriptions" WHERE "katello_subscriptions"."organization_id" = $1 AND "katello_subscriptions"."id" = $2 LIMIT $3  [["organization_id", 447626438], ["id", 904941504], ["LIMIT", 1]]
``` 

The code is organized so that viewing subscriptions (actually, they are Katello::Pool) use the Katello::Subscription permissions, at least for reading.

Reply:
> There are a _lot_ of negations going on in this code and it's hurting my brain. 
> ```ruby
> unless User.current.admin? || permitted_org_ids.include?(@resource.organization_id)
>   # raise error
> end
> ```
> 
> Now I can read the logic the same way I would describe the intended behavior: unless the user is an admin or subscription's org is permitted then raise an error.

I was thinking the same thing but couldn't think of a way to make it more clear with an `unless`!  jturel++